### DISCUSSION TITLE
how to get oauth token?

### DISCUSSION THREAD

**Discussion by yadavsingh** [spatie/laravel-google-calendar#177] [Q&A] [answered]
URL: https://developers.google.com/calendar/quickstart/php
This is generating only credentials.json, not token.json. How to get token.json?

**Comment by jl0nd0n0** [comment]
Run sample according to the instructions

In this line generate file token.json:

file_put_contents($tokenPath, json_encode($client->getAccessToken()));

**Comment by yadavsingh** [comment]
Thanks for your reply:

what i need to do here: google-calendar.php
'token_json' => storage_path('app/google-calendar/oauth-token.json'),

where i need that code set?
file_put_contents($tokenPath, json_encode($client->getAccessToken()));

**Comment by AntoineGUENEAU** [comment, marked as answer]
It's already set in quickstart.php, juste before the end : 

```
<?php
require __DIR__ . '/vendor/autoload.php';

if (php_sapi_name() != 'cli') {
    throw new Exception('This application must be run on the command line.');
}

/**
 * Returns an authorized API client.
 * @return Google_Client the authorized client object
 */
function getClient()
{
    $client = new Google_Client();
    $client->setApplicationName('Google Calendar API PHP Quickstart');
    $client->setScopes(Google_Service_Calendar::CALENDAR);
    $client->setAuthConfig('credentials.json');
    $client->setAccessType('offline');
    $client->setPrompt('select_account consent');

    // Load previously authorized token from a file, if it exists.
    // The file token.json stores the user's access and refresh tokens, and is
    // created automatically when the authorization flow completes for the first
    // time.
    $tokenPath = 'token.json';
    if (file_exists($tokenPath)) {
        $accessToken = json_decode(file_get_contents($tokenPath), true);
        $client->setAccessToken($accessToken);
    }

    // If there is no previous token or it's expired.
    if ($client->isAccessTokenExpired()) {
        // Refresh the token if possible, else fetch a new one.
        if ($client->getRefreshToken()) {
            $client->fetchAccessTokenWithRefreshToken($client->getRefreshToken());
        } else {
            // Request authorization from the user.
            $authUrl = $client->createAuthUrl();
            printf("Open the following link in your browser:\n%s\n", $authUrl);
            print 'Enter verification code: ';
            $authCode = trim(fgets(STDIN));

            // Exchange authorization code for an access token.
            $accessToken = $client->fetchAccessTokenWithAuthCode($authCode);
            $client->setAccessToken($accessToken);

            // Check to see if there was an error.
            if (array_key_exists('error', $accessToken)) {
                throw new Exception(join(', ', $accessToken));
            }
        }
        // Save the token to a file.
        if (!file_exists(dirname($tokenPath))) {
            mkdir(dirname($tokenPath), 0700, true);
        }
       file_put_contents($tokenPath, json_encode($client->getAccessToken())); //HERE
    }
    return $client;
}

```

**Reply by navaneeth-raman-bhaskar** [reply]
> It's already set in quickstart.php, juste before the end :
> 
> ```
> <?php
> require __DIR__ . '/vendor/autoload.php';
> 
> if (php_sapi_name() != 'cli') {
>     throw new Exception('This application must be run on the command line.');
> }
> 
> /**
>  * Returns an authorized API client.
>  * @return Google_Client the authorized client object
>  */
> function getClient()
> {
>     $client = new Google_Client();
>     $client->setApplicationName('Google Calendar API PHP Quickstart');
>     $client->setScopes(Google_Service_Calendar::CALENDAR);
>     $client->setAuthConfig('credentials.json');
>     $client->setAccessType('offline');
>     $client->setPrompt('select_account consent');
> 
>     // Load previously authorized token from a file, if it exists.
>     // The file token.json stores the user's access and refresh tokens, and is
>     // created automatically when the authorization flow completes for the first
>     // time.
>     $tokenPath = 'token.json';
>     if (file_exists($tokenPath)) {
>         $accessToken = json_decode(file_get_contents($tokenPath), true);
>         $client->setAccessToken($accessToken);
>     }
> 
>     // If there is no previous token or it's expired.
>     if ($client->isAccessTokenExpired()) {
>         // Refresh the token if possible, else fetch a new one.
>         if ($client->getRefreshToken()) {
>             $client->fetchAccessTokenWithRefreshToken($client->getRefreshToken());
>         } else {
>             // Request authorization from the user.
>             $authUrl = $client->createAuthUrl();
>             printf("Open the following link in your browser:\n%s\n", $authUrl);
>             print 'Enter verification code: ';
>             $authCode = trim(fgets(STDIN));
> 
>             // Exchange authorization code for an access token.
>             $accessToken = $client->fetchAccessTokenWithAuthCode($authCode);
>             $client->setAccessToken($accessToken);
> 
>             // Check to see if there was an error.
>             if (array_key_exists('error', $accessToken)) {
>                 throw new Exception(join(', ', $accessToken));
>             }
>         }
>         // Save the token to a file.
>         if (!file_exists(dirname($tokenPath))) {
>             mkdir(dirname($tokenPath), 0700, true);
>         }
>        file_put_contents($tokenPath, json_encode($client->getAccessToken())); //HERE
>     }
>     return $client;
> }
> ```

Generated token will expire as time pass by. How to refresh token?
